src/type_tracer/analyzer.rs

```diff
@@ -0,0 +1,1660 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::collections::HashSet;
+
+use anyhow::Result;
+use deno_ast::swc::ast::*;
+use deno_ast::swc::common::comments::CommentKind;
+use deno_ast::swc::utils::find_pat_ids;
+use deno_ast::swc::visit::*;
+use deno_ast::ModuleSpecifier;
+use deno_ast::ParsedSource;
+use deno_ast::SourcePos;
+use deno_ast::SourceRange;
+use deno_ast::SourceRangedForSpanned;
+use indexmap::IndexMap;
+use indexmap::IndexSet;
+
+use crate::CapturingModuleParser;
+use crate::ModuleGraph;
+use crate::ModuleParser;
+
+use super::collections::AdditiveOnlyMap;
+use super::cross_module;
+use super::cross_module::Definition;
+use super::TypeTraceDiagnostic;
+use super::TypeTraceDiagnosticKind;
+use super::TypeTraceHandler;
+
+#[derive(Debug, Clone)]
+pub struct RootSymbol {
+  specifiers_to_ids: HashMap<ModuleSpecifier, ModuleId>,
+  ids_to_symbols: HashMap<ModuleId, ModuleSymbol>,
+}
+
+impl RootSymbol {
+  pub fn get_module_from_specifier(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Option<&ModuleSymbol> {
+    let id = self.specifiers_to_ids.get(specifier)?;
+    self.ids_to_symbols.get(id)
+  }
+
+  pub fn get_module_from_id(
+    &self,
+    module_id: ModuleId,
+  ) -> Option<&ModuleSymbol> {
+    self.ids_to_symbols.get(&module_id)
+  }
+
+  pub fn into_specifier_map(self) -> IndexMap<ModuleSpecifier, ModuleSymbol> {
+    use std::collections::BTreeMap;
+
+    let ids_to_symbols =
+      self.ids_to_symbols.into_iter().collect::<BTreeMap<_, _>>();
+    let mut result = IndexMap::default();
+    for (id, symbol) in ids_to_symbols {
+      // todo(dsherret): improve
+      result.insert(
+        self
```

Comment:
Do you want to do that before landing?

Reply:
I think it's just a performance concern. I'll fix it later.